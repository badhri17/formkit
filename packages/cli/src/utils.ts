import { readdir, readFile, writeFile } from 'fs/promises'
import { resolve } from 'pathe'
import { exec } from 'child_process'
import { promisify } from 'util'

/**
 * Checks if a given directory is empty.
 * @param path - The absolute path to check if empty.
 */
export async function isDirEmpty(path: string) {
  try {
    const entries = await readdir(path)
    return entries.length === 0
  } catch (error) {
    return true
  }
}

export async function readPackageJSON(dir: string) {
  const packageJsonPath = resolve(dir, `./package.json`)
  const raw = await readFile(packageJsonPath, 'utf-8')
  return JSON.parse(raw)
}

export async function writePackageJSON(dir: string, data: any) {
  const packageJsonPath = resolve(dir, `./package.json`)
  await writeFile(packageJsonPath, JSON.stringify(data, null, 2))
}

const execAsync = promisify(exec)

// Async function to get the Git user name and email
export const getGitUser = async () => {
  const { stdout: name } = await execAsync('git config --get user.name')
  return name.trim()
}
