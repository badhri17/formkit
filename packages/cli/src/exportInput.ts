import axios from 'axios'
import { FORMKIT_VERSION } from '@formkit/core'
import { inputs } from '@formkit/inputs'
import { token } from '@formkit/utils'
import { resolve, relative, isAbsolute } from 'path'
import { access, mkdir, writeFile, readFile } from 'fs/promises'
import { error, info, warning, green, __dirname } from './index'
import { constants, existsSync } from 'fs'
import prompts from 'prompts'

/**
 * @internal
 */
export async function exportInput(
  inputOption?: string,
  options: Record<string, string | undefined> = {}
): Promise<void> {
  const input = await requireInput(inputOption)
  const lang = await requireLang(options.lang)
  const exportData = await requireInputCode(input, lang)
  const sourceCode = transformSource(exportData, input)
  const [absoluteDir, relativeDir] = await requireOutputDir(options.dir)
  const validDir = await upsertDir(absoluteDir)

  if (validDir === false)
    return error(`${relativeDir} is not a writable directory.`)
  if (!validDir) return
  const outFile = resolve(absoluteDir, `${input}.${lang}`)
  if (!existsSync(outFile)) {
    await writeFile(outFile, sourceCode)
  } else {
    return error('Did not export input because that file already exists.')
  }
  green(`Success! Exported ${relativeDir}/${input}.${lang}`)
  console.log(`To use it pass it to your FormKit configuration:
  // ...
  import { ${input} } from '${relativeDir}/${input}'
  // ...
  const config = defaultConfig({
    inputs: {
      ${input}
    }
  })
  `)
}

/**
 * Checks if a given directory is writable, if it doesn't exist, create it.
 * @param dir - A directory to create if it doesn't exist.
 * @returns
 */
async function upsertDir(dir: string): Promise<boolean | void> {
  if (!existsSync(dir)) {
    const local = '.' + dir.replace(process.cwd(), '')
    const { confirm } = await prompts({
      type: 'confirm',
      name: 'confirm',
      initial: true,
      message: `Export directory does not exist (${local}) does not exist. Create it?`,
    })
    if (!confirm) return info('Directory not created — no input was exported.')
    try {
      await mkdir(dir, { recursive: true })
    } catch {
      return error(`Unable to create directory ${dir}.`)
    }
  }
  try {
    await access(dir, constants.W_OK)
    return true
  } catch (err) {
    return false
  }
}

/**
 * Attempts to intelligently determine the directory to export to.
 * @param dir - The directory to export the input to.
 * @returns
 */
async function requireOutputDir(
  dir?: string
): Promise<[string, string]> | never {
  if (dir && isAbsolute(dir)) {
    return [dir, relative(process.cwd(), dir)]
  } else if (dir) {
    const abs = resolve(process.cwd(), dir)
    return [abs, relative(process.cwd(), abs)]
  }
  const rel = await prompts({
    type: 'text',
    name: 'dir',
    message:
      'Where should the input be exported to (relative to the current directory)?',
    initial: guessDir(),
  }).then((res) => res.dir)
  const abs = resolve(process.cwd(), rel)
  return [abs, rel]
}

function guessDir() {
  const srcDir = resolve(process.cwd(), 'src')
  if (existsSync(srcDir)) {
    return './src/inputs'
  }
  return './inputs'
}

/**
 * Transforms the source code of the inputs to be used locally.
 * @param exportData - The code to export.
 * @returns
 */
function transformSource(exportData: string, type: string): string | never {
  if (exportData) {
    // Change the exports from relative to npm package based.
    exportData = exportData.replace(
      /(}\sfrom\s['"])\.\.\/(?:index(?:\.mjs)?)?(['"])?/g,
      '$1@formkit/inputs$2'
    )
    const memoKey = token()
    exportData = exportData.replace(
      /(schemaMemoKey:\s?['"])[a-zA-Z0-9]+(['"])/g,
      `$1${memoKey}$2`
    )
    // Inject the forceTypeProp in the definition.
    exportData = exportData.replace(
      /^  props: \[(.*)\],/gm,
      `  props: [$1],
  /**
   * Forces node.props.type to be this explicit value.
   */
  forceTypeProp: '${type}',`
    )
  } else {
    error('Unable to export the input file because it cannot be located.')
  }
  return exportData
}

/**
 * Determine the language the user wants to export.
 * @param lang - The language to export the input to.
 * @returns
 */
async function requireLang(lang?: string): Promise<string> {
  if (!lang) {
    const guessedLang = guessLang()
    lang = await prompts({
      type: 'select',
      name: 'lang',
      message: 'What language should be used?',
      choices: [
        guessedLang === 'ts'
          ? { title: 'TypeScript', value: 'ts' }
          : { title: 'JavaScript', value: 'js' },
        guessedLang === 'ts'
          ? { title: 'JavaScript', value: 'js' }
          : { title: 'TypeScript', value: 'ts' },
      ],
    }).then((val) => val.lang)
    if (!lang) {
      error('No language selected, exiting.')
    }
  }
  return lang
}

/**
 * Fetch the input name that the user wants to export.
 * @param inputName - The name of the input to load.
 * @returns
 */
async function requireInput(inputName?: string): Promise<string> | never {
  if (!inputName) {
    const res = await prompts({
      type: 'autocomplete',
      name: 'inputName',
      message: 'What input do you want to export?',
      choices: Object.keys(inputs).map((i) => ({
        title: i,
        value: i,
      })),
    })
    inputName = res.inputName
  }
  if (!inputName || !(inputName in inputs)) {
    error(
      `Cannot export “${inputName}” because it is not part of the @formkit/inputs package.`
    )
  }
  return inputName
}

/**
 * Loads the string data of an input that should be exported.
 * @param name - The name of the input to load.
 * @param lang - The language to load the input in.
 * @returns
 */
async function requireInputCode(
  name: string,
  lang?: string
): Promise<string> | never {
  lang = !lang ? guessLang() : lang
  const localFile = resolve(
    __dirname,
    `../../inputs/dist/exports/${name}.${lang}`
  )
  let fileData = null
  if (existsSync(localFile)) {
    fileData = await readFile(localFile, { encoding: 'utf8' })
  } else {
    warning(`Unable to locate ${localFile}`)
    const cdnUrl = `https://cdn.jsdelivr.net/npm/@formkit/inputs@${FORMKIT_VERSION}/dist/exports/${name}.${lang}`
    try {
      const res = await axios.get(cdnUrl)
      fileData = res.data
    } catch (e: any) {
      if (e && e?.response?.status) {
        error(`${e.response.status} — unable to load ${localFile}`)
      } else {
        error(
          'Unable to load input file — probably a network error. Are you online?'
        )
      }
    }
  }

  if (!fileData) {
    info('Checking CDN for an exportable input.')
  }

  if (!fileData) {
    error(`Unable to load export ${name}.${lang}`)
  } else {
    return fileData
  }
}

/**
 * Guess the language the user is leveraging on their project.
 */
function guessLang() {
  const tsconfig = resolve(process.cwd(), 'tsconfig.json')
  return existsSync(tsconfig) ? 'ts' : 'js'
}
