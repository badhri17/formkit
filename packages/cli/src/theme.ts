import { existsSync } from 'fs'
import http from 'http'
import { writeFile, readFile } from 'fs/promises'
import { resolve } from 'pathe'
import { error, red, green, __dirname, info } from './index'
import { Theme, ThemeOptions } from '@formkit/theme-creator'
import { stylesheetFromTailwind } from '@formkit/theme-creator/stylesheet'
import { slugify } from '@formkit/utils'
import prompts from 'prompts'
import { createHash } from 'crypto'
import ora, { Ora } from 'ora'
import open from 'open'
import { parse as parseUrl } from 'url'
import { token } from '@formkit/utils'
import { getPort } from 'get-port-please'

interface BuildThemeOptions {
  semantic: boolean
  theme?: string
  variables?: string
  api?: string
  format?: 'ts' | 'mjs'
  outFile?: string
}

interface ServerResponse {
  status: 'success' | 'failed'
  message: string
}

const DEFAULT_THEME_API = 'https://themes.formkit.com/api'
const DEFAULT_THEME_EDITOR = 'https://themes.formkit.com'
const HAS_EXTENSION_RE = /\.(?:ts|js|mjs|cjs)$/

let themeListResponse: Response | null = null
let themes: Array<Record<string, any>> = []

async function fetchThemes() {
  themeListResponse = await fetch(`${DEFAULT_THEME_API}/themes`)
  themes = await themeListResponse.json()
}

export async function buildTheme(options: Partial<BuildThemeOptions> = {}) {
  if (!themes.length) await fetchThemes()

  let themeName = options.theme || ''

  if (!options.theme) {
    const generatedTheme = await localGeneratedTheme()

    if (generatedTheme) {
      const [code, path] = generatedTheme
      try {
        const [checksum, variables, theme] = extractThemeData(code)

        // determine if the theme is from themes.formkit.com
        const themeFromList = themes.find((t: any) => t.slug === theme)

        if (themeFromList) {
          const { editMode } = await prompts({
            type: 'confirm',
            message: `Found local theme file for ${theme}, edit this theme?`,
            name: 'editMode',
            initial: true,
          })
          if (editMode) {
            return await editTheme(path, code, checksum, variables, theme)
          }
        } else if (generatedTheme) {
          red(
            'It appears you have a local theme installed that is not part of the themes.formkit.com registry. By continuing your local theme file will be overwritten.'
          )
          const { confirm } = await prompts({
            type: 'confirm',
            message: 'Are you sure you want to continue?',
            name: 'confirm',
            initial: false,
          })

          if (!confirm) {
            return error('Aborting.')
          }
        }
      } catch (err) {
        console.error(err)
        console.info(
          `Detected local theme file but could not parse it. Perhaps it was edited or not generated by FormKit?`
        )
      }
    }
    const { theme } = await prompts({
      type: 'select',
      message: 'Select a theme',
      name: 'theme',
      choices: themes.map((theme: any) => ({
        title: theme.name,
        value: theme.slug,
        description:
          theme.description +
          (theme.darkMode ? ' 🌜' : '') +
          (theme.lightMode ? '☀️' : '') +
          ')',
      })),
    })
    themeName = theme
  }
  if (!themeName) error('Please provide a theme name or path to a theme file.')
  green(`Locating ${themeName}...`)

  const format = options.format || guessFormat()

  let theme = null
  // if the theme is in the themesList then grab it from the api
  const themeFromList = themes.find((t: any) => t.slug === themeName)

  // check if there is a local file matching the provided theme name
  const themeFilePath = determineFilePath(themeName) || themeName
  const localThemeFile = existsSync(resolve(process.cwd(), themeFilePath))
    ? await readFile(themeName, 'utf-8')
    : null

  if (themeFromList) {
    info('Fetching theme from themes.formkit.com')
    theme = await apiTheme(
      themeName,
      DEFAULT_THEME_API,
      options.variables ?? '',
      format === 'ts',
      options.semantic ?? false
    )
  } else if (localThemeFile) {
    info('Found local theme file')
    theme = await generate(
      await localTheme(themeName),
      options.variables,
      format === 'ts',
      options.semantic,
      themeName
    )
  } else {
    // try to load theme from locally installed package
    // using dynamic import
    info('Fetching theme from local npm packages')
    try {
      const { default: themeFunction } = await import(themeName)
      theme = await generate(
        themeFunction,
        options.variables,
        format === 'ts',
        options.semantic,
        themeName
      )
      green(`Found locally installed theme package: ${themeName}`)
    } catch (err) {
      console.error(err)
    }
  }

  if (theme) {
    const outFile =
      options.outFile || 'formkit.theme.' + (options.semantic ? 'css' : format)
    await writeFile(resolve(process.cwd(), outFile), theme)
    green(`Theme file written to ${outFile}`)
  } else {
    red(`Could not find theme: ${theme}`)
  }
}

async function editTheme(
  path: string,
  code: string,
  checksum: string,
  variables: string,
  theme: string
) {
  const codeWithoutChecksum = code.replace(
    `* @checksum - ${checksum}`,
    '* @checksum -'
  )
  const newChecksum = createHash('sha256')
    .update(codeWithoutChecksum)
    .digest('hex')
  if (newChecksum !== checksum) {
    const { confirm } = await prompts({
      type: 'confirm',
      message:
        'It appears you’ve edited your theme file. By continuing any changes made to your theme file will be lost. Are you sure you want to continue?',
      name: 'confirm',
      initial: false,
    })

    if (!confirm) {
      return error('Aborting.')
    }
  }
  await editMode(theme, variables, path)
}

/**
 * Use the public formkit url shortener.
 * @param url - A url to shorten
 */
async function shorten(url: string): Promise<string> {
  const res = await fetch('https://www.formk.it', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ url }),
  })
  const data = await res.json()
  return data.url
}

async function editMode(
  theme: string,
  variables: string,
  filePath: string
): Promise<void> {
  const nonce = token()
  const port = await getPort({
    portRange: [5480, 5550],
    host: 'localhost',
  })
  const url = `${DEFAULT_THEME_EDITOR}/editor/?theme=${theme}&variables=${variables}&n=${nonce}&p=${port}`
  const shortUrl = await shorten(url)
  info(`Open the theme editor at: ${shortUrl}`)
  const spinner = ora(`Edit in your browser.`).start()
  await open(url)
  let heartBeatTimeout: NodeJS.Timeout | undefined
  let server: http.Server | undefined
  let completedSuccessfully = false
  let savedAt = 0
  await new Promise<void>((resolve) => {
    server = http
      .createServer(async (req, res) => {
        const urlObj = parseUrl(req.url!, true)
        const path = urlObj.pathname as string
        const theme = urlObj.query.theme as string | undefined
        const variables = urlObj.query.variables as string | undefined
        const resData: ServerResponse = {
          status: 'failed',
          message: 'Incorrect token.',
        }
        if (path === `/${nonce}`) {
          clearTimeout(heartBeatTimeout)
          resData.status = 'success'
          if (theme && typeof variables !== 'undefined') {
            resData.message = 'updated'
            const themeCode = await apiTheme(
              theme,
              DEFAULT_THEME_API,
              variables,
              filePath.endsWith('.ts'),
              false,
              spinner
            )
            await writeFile(filePath, themeCode)
            spinner.text = `⚡️ theme updated.`
            savedAt = Date.now()
          } else {
            resData.message = 'listening'
            if (Date.now() - savedAt > 5000) {
              spinner.text = `waiting for${savedAt ? ' more' : ''} changes...`
            }
            heartBeatTimeout = setTimeout(() => {
              completedSuccessfully = true
              resolve()
            }, 3000)
          }
        }
        res.writeHead(resData.status === 'success' ? 200 : 400, {
          'Content-Type': 'application/json',
          'Access-Control-Allow-Origin': '*',
          Vary: 'Origin',
        })
        res.end(JSON.stringify(resData))
      })
      .listen(port)
  })
  server?.close()
  spinner.stop()
  if (completedSuccessfully) {
    info('Theme editor session complete, enjoy your theme 🎨')
    process.exit(0)
  }
}

function guessFormat() {
  return existsSync(resolve(process.cwd(), 'tsconfig.json')) ? 'ts' : 'mjs'
}

export async function generate(
  theme: Theme<ThemeOptions> | ReturnType<Theme<ThemeOptions>>,
  variables?: string,
  isTS?: boolean,
  semantic?: boolean,
  themeName?: string
): Promise<string> {
  if (typeof theme === 'function') {
    const vars = parseVariables(variables)
    themeName = theme.meta.name
    theme = theme(vars)
  }
  if (!themeName) error('Could not determine theme name.')
  info(`Loaded theme: ${themeName}`)
  const classes = theme.tailwind()
  if (semantic) return await stylesheetFromTailwind(classes)
  const classList: Record<string, Record<string, true>> = {}
  const globals: Record<string, Record<string, true>> = {}

  for (const input in classes) {
    for (const section in classes[input]) {
      const key = `${input}__${section}`
      const sectionClasses = classes[input][section]
        .split(' ')
        .reduce((acc, cur) => {
          acc[cur] = true
          return acc
        }, {} as Record<string, true>)
      if (input === '__globals') {
        globals[section] = sectionClasses
      } else {
        classList[key] = sectionClasses
      }
    }
  }

  const themeFile = `${
    isTS ? "import type { FormKitNode } from '@formkit/core'\n\n" : ''
  }/**
  * @privateRemarks
  * This file was generated by the FormKit CLI and should not be manually
  * edited unless you’d like to "eject" from the CLI’s ability to update it.
  *
  * @checksum -
  * @variables - ${variables}
  * @theme - ${slugify(themeName)}
  **/

 /**
  * This is the theme function itself, it should be imported and used as the
  * config.rootClasses function. For example:
  *
  * \`\`\`js
  * import { theme } from './formkit.theme'
  * import { defineFormKitConfig } from '@formkit/vue'
  *
  * export default defineFormKitConfig({
  *   config: {
  *     rootClasses: theme
  *   }
  * })
  * \`\`\`
  **/
 export function rootClasses (sectionName${isTS ? ': string' : ''}, node${
    isTS ? ': FormKitNode' : ''
  })${isTS ? ': Record<string, boolean>' : ''} {
   const key = \`\${node.props.type}__\${sectionName}\`
   const semanticKey = \`formkit-\${sectionName}\`
   const familyKey = node.props.family ? \`family:\${node.props.family}__\${sectionName}\` : ''
   const memoKey = \`\${key}__\${familyKey}\`
   if (!(memoKey in classes)) {
     const sectionClasses = classes[key] ?? globals[sectionName] ?? {}
     sectionClasses[semanticKey] = true
     if (familyKey in classes) {
       classes[memoKey] = { ...classes[familyKey],  ...sectionClasses }
     } else {
       classes[memoKey] = sectionClasses
     }
   }
   return classes[memoKey] ?? { [semanticKey]: true }
 }

/**
 * These classes have already been merged with globals using tailwind-merge
 * and are ready to be used directly in the theme.
 **/
const classes${
    isTS ? ': Record<string, Record<string, boolean>>' : ''
  } = ${JSON.stringify(classList, null, 2)};

/**
 * Globals are merged prior to generating this file — these are included for
 * any other non-matching inputs.
 **/
const globals${
    isTS ? ': Record<string, Record<string, boolean>>' : ''
  } = ${JSON.stringify(globals, null, 2)};
`

  const checksum = createHash('sha256').update(themeFile).digest('hex')
  return themeFile.replace(/@checksum -/, `@checksum - ${checksum}`)
}

function parseVariables(variables?: string): Record<string, string> {
  if (!variables) return {}
  return variables.split(',').reduce((vars, unparsed) => {
    const [key, value] = unparsed.split('=')
    vars[key] = value
    return vars
  }, {} as Record<string, string>)
}

function determineFilePath(fileName: string): string | undefined {
  const extensions = ['.ts', '.js', '.mjs', '.cjs']
  const paths = extensions.map((ext) => resolve(process.cwd(), fileName + ext))
  return getPath(paths)
}

function getPath(paths: string[]): string | undefined {
  const path = paths.shift()
  if (existsSync(path!)) return path
  return paths.length ? getPath(paths) : undefined
}

async function localGeneratedTheme(): Promise<[string, string] | undefined> {
  const path = determineFilePath('formkit.theme')
  if (path) {
    const code = await readFile(path, 'utf-8')
    return [code, path]
  }
  return undefined
}

async function localTheme(
  themeName: string
): Promise<Theme<ThemeOptions>> | never {
  const extensions = ['.ts', '.js', '.mjs', '.cjs']
  const paths = HAS_EXTENSION_RE.test(themeName)
    ? [resolve(process.cwd(), themeName)]
    : extensions.map((ext) => resolve(process.cwd(), themeName + ext))
  const path = getPath(paths)
  if (!path) error(`Could not find ${themeName}.`)

  const theme = (await import(path)) as { default: Theme<ThemeOptions> }
  if (typeof theme !== 'object' || !theme.default) error('Invalid theme file.')
  return theme.default
}

export function extractThemeData(
  theme: string
): [string, string, string] | never {
  const checksumStart = theme.indexOf('@checksum -')
  const variablesStart = theme.indexOf('@variables -')
  const themeStart = theme.indexOf('@theme -')
  if (checksumStart === -1 || variablesStart === -1 || themeStart === -1) {
    throw new Error('Unable to find checksum in theme file.')
  }
  const checksum = theme.substring(
    checksumStart + 12,
    theme.indexOf('\n', checksumStart)
  ) as string
  const variables = theme.substring(
    variablesStart + 13,
    theme.indexOf('\n', variablesStart)
  ) as string
  const themeName = theme.substring(
    themeStart + 9,
    theme.indexOf('\n', themeStart)
  ) as string
  return [checksum, variables, themeName]
}

async function apiTheme(
  themeName: string,
  endpoint: string,
  variables: string,
  isTS: boolean,
  semantic: boolean,
  spinner?: Ora
): Promise<string> {
  if (spinner) {
    spinner.text = `🔨 building ${themeName}`
  } else {
    info(`⚡️ building theme: ${themeName}`)
  }
  const res = await fetch(`${endpoint}/generate`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      theme: themeName,
      variables,
      ts: String(isTS),
      semantic: String(semantic),
    }),
  })
  if (res.ok) {
    const code = await res.text()
    return code
  } else {
    error(`Could not generate theme — ${res.statusText}`)
  }
}
