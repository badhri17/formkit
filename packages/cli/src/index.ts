import { FORMKIT_VERSION } from '@formkit/core'
import { Command } from 'commander'
import chalk from 'chalk'
import { exportInput } from './exportInput'
import { createApp } from './createApp'
import { buildTheme } from './theme'
import { createTheme } from './createTheme'
import { dirname } from 'path'
import { fileURLToPath } from 'url'
import { execa } from 'execa'

const __filename = fileURLToPath(import.meta.url)

/**
 * @internal
 */
export const __dirname = dirname(__filename)

/**
 * @internal
 */
export const red = (m: string): void =>
  console.log(`${chalk.red.bold('[FormKit Error]: ')} ${chalk.red(m)}`)

/**
 * @internal
 */
export const info = (m: string): void => {
  console.log(`${chalk.blue(m)}`)
}

/**
 * @internal
 */
export const warning = (m: string): void => {
  console.log(`${chalk.yellow.bold('[FormKit Warn]: ')} ${chalk.yellow(m)}`)
}

/**
 * @internal
 */
export const green = (m: string): void => {
  console.log(chalk.greenBright(m))
}

const program = new Command()

program
  .name('FormKit CLI')
  .description('The official FormKit command line utility.')
  .version(FORMKIT_VERSION)

program
  .command('export')
  .argument(
    '[input]',
    'An input to export (from @formkit/inputs, like "text" or "select")'
  )
  .description(
    'Export an input from @formkit/inputs to a local file for modification.'
  )
  .option('-d, --dir <dir>', 'The directory to export inputs to')
  .option('-l, --lang <ts|js>', 'Export as TypeScript (ts) or JavaScript (js)')
  .action(exportInput)

program
  .command('create-app')
  .argument('[name]', 'Creates a vite application with this name.')
  .option(
    '-l, --lang <ts|js>',
    'Create a TypeScript (ts) or JavaScript (js) app.'
  )
  .option('-f, --framework <vite|nuxt>', 'Create a new Vite or Nuxt app.')
  .option(
    '--pro <key>',
    'Creates the project with FormKit pro installed if a project key is provided.'
  )
  .description('Creates a new Vue 3 Vite application with FormKit installed.')
  .action(createApp)

program
  .command('theme')
  .option(
    '--theme <theme>',
    'A published theme or local npm theme package identifier.'
  )
  .option(
    '-s',
    '--semantic',
    'Build the theme as semantic classes rather than utility classes.'
  )
  .option(
    '--variables <variables>',
    'A comma separated list of variables to include in the theme. For example: --variables "primaryColor=#221233, padding=10"'
  )
  .option(
    '--api <url>',
    'An alternative API endpoint to use for theme fetching and generation.'
  )
  .description(
    'Creates a FormKit theme file (formkit.theme.ts) using a known theme and variables.'
  )
  .action(buildTheme)

program
  .command('create-theme')
  .argument('[name]', 'The public name of the theme, for example "Monokai".')
  .option(
    '--package-name <packageName>',
    'The name of the package on npm, for example formkit-theme-monokai.'
  )
  .option('--dir <directory>', 'The relative directory to create the theme in.')
  .description('Scaffold a new theme from @formkit/theme-starter.')
  .action(createTheme)

/**
 * @internal
 */
export async function cli(): Promise<void> {
  const res = await execa('npx', ['--version'])
  const [major] = res.stdout.trim().split('.')
  if (Number(major) < 7) {
    error(`Your npm version must be 7 or higher (found ${res.stdout.trim()}).`)
  }
  program.parse()
}

export * from './theme'

/**
 * @internal
 */
export function error(message: string): never {
  red(message)
  process.exit(1)
}
