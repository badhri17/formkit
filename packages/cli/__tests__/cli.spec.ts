import { describe, it, expect, vi, beforeAll } from 'vitest'
import { buildTheme, extractThemeData } from '../src/theme'
import chalk from 'chalk'
import { readFile } from 'fs/promises'
import { mkdirSync } from 'fs'
import { resolve } from 'pathe'
import { createNode } from '@formkit/core'

beforeAll(() => {
  mkdirSync(resolve(process.cwd(), 'temp'), { recursive: true })
})

describe('buildTheme', () => {
  it('can generate a local theme', async () => {
    const consoleMock = vi
      .spyOn(console, 'log')
      .mockImplementation(() => undefined)
    await buildTheme({
      theme: './packages/cli/__tests__/mocks/localTheme.ts',
      outFile: 'temp/formkit.theme.ts',
      format: 'ts',
    })

    expect(consoleMock).toHaveBeenNthCalledWith(
      1,
      chalk.greenBright(
        'Locating ./packages/cli/__tests__/mocks/localTheme.ts...'
      )
    )
    const fileString = await readFile(
      resolve(process.cwd(), 'temp/formkit.theme.ts'),
      'utf-8'
    )
    expect(fileString).toMatchSnapshot()
  })

  it('can override variables in generated theme', async () => {
    await buildTheme({
      theme: './packages/cli/__tests__/mocks/localTheme.ts',
      outFile: 'temp/formkit.theme.ts',
      format: 'ts',
      variables: 'border=border-6,spacing=10',
    })
    const fileString = await readFile(
      resolve(process.cwd(), 'temp/formkit.theme.ts'),
      'utf-8'
    )
    expect(fileString).toMatchSnapshot()
  })

  it('returns the correct classes for a given input section', async () => {
    await buildTheme({
      theme: './packages/cli/__tests__/mocks/localTheme.ts',
      outFile: 'temp/formkit.theme.ts',
      format: 'ts',
      variables: 'spacing=5',
    })
    // @ts-ignore
    const { rootClasses } = await import(
      resolve(process.cwd(), 'temp/formkit.theme.ts')
    )
    const node = createNode({
      type: 'input',
      props: { type: 'text', rootClasses, family: 'text' },
    })

    // @ts-ignore
    expect(node.props.rootClasses!('outer', node)).toEqual({
      'border-green-300': true,
      'formkit-outer': true,
      'mb-5': true,
      'ml-80': true,
      'mr-10': true,
      'mt-2': true,
      'text-green-300': true,
    })
  })
})

describe('extractThemeData', () => {
  it('can extract basic details about a given theme', async () => {
    await buildTheme({
      theme: './packages/cli/__tests__/mocks/localTheme.ts',
      outFile: 'temp/formkit.theme.ts',
      format: 'ts',
      variables: 'spacing=5',
    })
    const fileString = await readFile(
      resolve(process.cwd(), 'temp/formkit.theme.ts'),
      'utf-8'
    )
    const themeData = extractThemeData(fileString)
    expect(themeData).toEqual([
      '9902ad56008296f4db97332febf34b6c1e82159dcb57f81f30198bca7c3ebe92',
      'spacing=5',
      'simple',
    ])
  })
})
