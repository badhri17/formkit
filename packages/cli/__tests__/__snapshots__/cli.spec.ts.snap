// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`buildTheme > can generate a local theme 1`] = `
"import type { FormKitNode } from '@formkit/core'

/**
  * @privateRemarks
  * This file was generated by the FormKit CLI and should not be manually
  * edited unless you’d like to \\"eject\\" from the CLI’s ability to update it.
  *
  * @checksum - 31339e326a2fa6bcd6a0a59923799f4a21092b07a24a3b3cfaf78b915e6558b2
  * @variables - undefined
  * @theme - simple
  **/

 /**
  * This is the theme function itself, it should be imported and used as the
  * config.rootClasses function. For example:
  *
  * \`\`\`js
  * import { theme } from './formkit.theme'
  * import { defineFormKitConfig } from '@formkit/vue'
  *
  * export default defineFormKitConfig({
  *   config: {
  *     rootClasses: theme
  *   }
  * })
  * \`\`\`
  **/
 export function rootClasses (sectionName: string, node: FormKitNode): Record<string, boolean> {
   const key = \`\${node.props.type}__\${sectionName}\`
   const semanticKey = \`formkit-\${sectionName}\`
   const familyKey = node.props.family ? \`family:\${node.props.family}__\${sectionName}\` : ''
   const memoKey = \`\${key}__\${familyKey}\`
   if (!(memoKey in classes)) {
     const sectionClasses = classes[key] ?? globals[sectionName] ?? {}
     sectionClasses[semanticKey] = true
     if (familyKey in classes) {
       classes[memoKey] = { ...classes[familyKey],  ...sectionClasses }
     } else {
       classes[memoKey] = sectionClasses
     }
   }
   return classes[memoKey] ?? { [semanticKey]: true }
 }

/**
 * These classes have already been merged with globals using tailwind-merge
 * and are ready to be used directly in the theme.
 **/
const classes: Record<string, Record<string, boolean>> = {
  \\"family:text__outer\\": {
    \\"mb-64\\": true,
    \\"mt-2\\": true,
    \\"ml-80\\": true,
    \\"mr-10\\": true,
    \\"border-green-300\\": true,
    \\"text-green-300\\": true
  },
  \\"text__input\\": {
    \\"border-2\\": true,
    \\"rounded-md\\": true,
    \\"p-2\\": true,
    \\"border-slate-50\\": true
  },
  \\"select__inner\\": {
    \\"border-2\\": true,
    \\"rounded-md\\": true,
    \\"p-4\\": true
  },
  \\"select__outer\\": {
    \\"mt-2\\": true,
    \\"ml-80\\": true,
    \\"mr-10\\": true,
    \\"border-green-300\\": true,
    \\"mb-56\\": true
  }
};

/**
 * Globals are merged prior to generating this file — these are included for
 * any other non-matching inputs.
 **/
const globals: Record<string, Record<string, boolean>> = {
  \\"outer\\": {
    \\"mb-64\\": true,
    \\"mt-2\\": true,
    \\"ml-80\\": true,
    \\"mr-10\\": true,
    \\"border-green-300\\": true
  }
};
"
`;

exports[`buildTheme > can override variables in generated theme 1`] = `
"import type { FormKitNode } from '@formkit/core'

/**
  * @privateRemarks
  * This file was generated by the FormKit CLI and should not be manually
  * edited unless you’d like to \\"eject\\" from the CLI’s ability to update it.
  *
  * @checksum - cdc3877bf8c21ade80f2ab0762482c4523279686adfd966d789abff9f80f2be1
  * @variables - border=border-6,spacing=10
  * @theme - simple
  **/

 /**
  * This is the theme function itself, it should be imported and used as the
  * config.rootClasses function. For example:
  *
  * \`\`\`js
  * import { theme } from './formkit.theme'
  * import { defineFormKitConfig } from '@formkit/vue'
  *
  * export default defineFormKitConfig({
  *   config: {
  *     rootClasses: theme
  *   }
  * })
  * \`\`\`
  **/
 export function rootClasses (sectionName: string, node: FormKitNode): Record<string, boolean> {
   const key = \`\${node.props.type}__\${sectionName}\`
   const semanticKey = \`formkit-\${sectionName}\`
   const familyKey = node.props.family ? \`family:\${node.props.family}__\${sectionName}\` : ''
   const memoKey = \`\${key}__\${familyKey}\`
   if (!(memoKey in classes)) {
     const sectionClasses = classes[key] ?? globals[sectionName] ?? {}
     sectionClasses[semanticKey] = true
     if (familyKey in classes) {
       classes[memoKey] = { ...classes[familyKey],  ...sectionClasses }
     } else {
       classes[memoKey] = sectionClasses
     }
   }
   return classes[memoKey] ?? { [semanticKey]: true }
 }

/**
 * These classes have already been merged with globals using tailwind-merge
 * and are ready to be used directly in the theme.
 **/
const classes: Record<string, Record<string, boolean>> = {
  \\"family:text__outer\\": {
    \\"mb-10\\": true,
    \\"mt-2\\": true,
    \\"ml-80\\": true,
    \\"mr-10\\": true,
    \\"border-green-300\\": true,
    \\"text-green-300\\": true
  },
  \\"text__input\\": {
    \\"border-6\\": true,
    \\"rounded-md\\": true,
    \\"p-2\\": true,
    \\"border-slate-50\\": true
  },
  \\"select__inner\\": {
    \\"border-6\\": true,
    \\"rounded-md\\": true,
    \\"p-4\\": true
  },
  \\"select__outer\\": {
    \\"mt-2\\": true,
    \\"ml-80\\": true,
    \\"mr-10\\": true,
    \\"border-green-300\\": true,
    \\"mb-8\\": true
  }
};

/**
 * Globals are merged prior to generating this file — these are included for
 * any other non-matching inputs.
 **/
const globals: Record<string, Record<string, boolean>> = {
  \\"outer\\": {
    \\"mb-10\\": true,
    \\"mt-2\\": true,
    \\"ml-80\\": true,
    \\"mr-10\\": true,
    \\"border-green-300\\": true
  }
};
"
`;
